### PR TITLE
Provide validators as a list in genesis file

### PR DESCRIPTION
**Pull request by JekaMas**
Nowadays the only way to set up validators is encoding those IDs into `extraData` Genesis field.
DevOps could exclude a few operations from their pipeline if it'd possible to provide validators in human-readable format in Genesis.
The user documentation would be much simpler too. A user would need only to edit Genesis and run `autonity init path_to_genesis`.

### CODE REVIEW

File: core/genesis.go [hunk 1]
```diff
@@ -260,6 +261,9 @@ func (g *Genesis) ToBlock(db ethdb.Database) *types.Block {
 	}
 	if g.Difficulty == nil {
 		head.Difficulty = params.GenesisDifficulty
+	}
+	if len(g.Validators) != 0 {
+
```
Comment (yazzaoui) [inline core/genesis.go]: Something is missing here

Comment (JekaMas) [reply core/genesis.go]: It was auto generated. Will check why go generate returned that code.

File: core/genesis.go [hunk 2]
```diff
@@ -314,6 +341,20 @@ func (g *Genesis) MustCommit(db ethdb.Database) *types.Block {
...
 }
 
+func (g *Genesis) UpdateValidators(addressList []string) error {
+	var validators []common.Address
+	for _, address := range addressList {
+		validators = append(validators, common.HexToAddress(address))
+	}
+
+	var err error
+	if g.ExtraData, err = types.PrepareExtra(&g.ExtraData, validators); err != nil {
```
Comment (yazzaoui) [inline core/genesis.go]: Why is this function taking a pointer on a slice as an argument and not the slice directly?

Comment (JekaMas) [reply core/genesis.go]: The method uses append function that can change the underlying struct of slice. So if we want slice to be changed after the method call, we have to use it by reference.
A small example https://play.golang.org/p/Ex5cZhzPLhg

Comment (yazzaoui) [reply core/genesis.go]: Oh yes, sorry forgot that, thank you for the example !

File: core/genesis.go (hunk 1 above)
Comment (JekaMas) [reply core/genesis.go]: I've checked that place. It'd a piece of dead code, the actual function was extracted into `setIstanbul`

File: core/network_permissioning.go [hunk 3]
```diff
@@ -163,7 +163,7 @@ func (bc *BlockChain) callGlienickeContract(state *state.StateDB, header *types.
 	}
 
 	var returnedEnodes []string
-	if err := ABI.Unpack(returnedEnodes, "getWhitelist", ret); err != nil { // can't work with aliased types
+	if err := ABI.Unpack(&returnedEnodes, "getWhitelist", ret); err != nil { // can't work with aliased types
```
Comment (yazzaoui) [inline core/network_permissioning.go]: Interesting, why this change?

**Comment by yazzaoui** [comment]
LGTM

File: core/genesis.go [hunk 4]
```diff
@@ -283,15 +297,44 @@ func (g *Genesis) Commit(db ethdb.Database) (*types.Block, error) {
...
+	if len(g.Validators) != 0 {
+		err := g.UpdateValidators(g.Validators)
+		if err != nil {
+			return fmt.Errorf("can't commit genesis block with incorrect validators: %s", err)
+		}
+	}
+	log.Info("starting Istanbul consensus", "extraData", common.Bytes2Hex(g.ExtraData))
+
+	// we have to use '1' to have TD == BlockNumber for xBFT consensus
+	g.Difficulty = big.NewInt(1)
```
Comment (aneequesafdar) [inline core/genesis.go]: Isn't the block number for genesis defined as 0? So why aren't we setting the difficulty to 1? Is it because difficulty 1 is inherited by block 1?

Comment (JekaMas) [reply core/genesis.go]: It is, however here we set `Difficulty` to 1, not the total difficulty. If we have `Difficulty==1` so for blockNumber==0 => TD=0, blockNumber==1 => TD=1, blockNumber==2 => TD=2 etc

IBFT relies on TD as equal to Block number, afaik.
